Fill in Overview.md (#1)
WebAssembly/module-types: proposals/module-types/Overview.md

# Module Types for WebAssembly

A proposal to enhance module types to include import/export names and define

> text format parse rules, so that module types can be written independently of

Reviewer: ```suggestion
text format syntax, so that module types can be written independently of
```

> module definition.

Reviewer: ```suggestion
module definitions.
```

2. assigning text format parse rules

For example, the module:

```wasm
(module
  (memory (import "a" "x") 1 2)
  (func (import "a" "y") (param i32))
  (table (export "b") 1 funcref)
  (func $notImportedOrExported (result i64)
    i64.const 0
  )
  (func (export "c") (result f32)
    f32.const 0
  )
)
```

would have a module type that could be written in the text format as:

```wasm
(module
  (memory (import "a" "x") 1 2)
  (func (import "a" "x") (param i32))
  (table (export "b") 1 funcref)
  (func (export "c") (result f32))
)
```

Here, the `module` keyword is used for both the definition and the type,
symmetric to the existing text format convention where, e.g., `func` is used
in both the definition and types of functions.

In WebAssembly there is also the separate concept of a module *instance*,
which is the result of [instantiating] a module with imports. In loader
frameworks like [ESM], where instantiation is performed by the loader,
not by client modules, or for built-in modules like [WASI], where instantiation
is logically performed by the host, the type of module *instances* may be
what is actually needed for the toolchain. Thus, the proposal also defines

> **instance type** along with text format parse rules. 

Reviewer: ```suggestion
**instance type** along with text format syntax. 
```

For example, the above module, when instantiated, would have an instance type
that could be written in the text format as:

```wasm
(instance
  (table (export "b") 1 funcref)
  (func (export "c") (result f32))
)
```

Just as the current [text format conventions recommend `.wat`][WAT] as the
extension of a file that contains a module definition, the proposal would
include a new recommendation for text files containing module types or
instance types.

> TODO: add a strawman recommendation (`.wmt`+`.wit`? or just share `.wit`?)

Reviewer: I like `.wit`, which is a wordplay similar to `.wat` but also reads like "witness". We can probably share it for both.

Author: Yeah, I was thinking the same thing (well, not literally "witness", but the sharing), but since I presented `.wmt`/`.wit`, I made this a TODO.  But if you agree, then maybe we should just propose `.wit`.
